src/librustc_trans/context.rs

```diff
@@ -967,6 +967,12 @@ impl<'b, 'tcx> CrateContext<'b, 'tcx> {
         name.push_str(prefix);
         name.push_str(".");
         base_n::push_str(idx as u64, base_n::MAX_BASE, &mut name);
+
+        // MSP430 assembler does not like '@' character in symbol names.
+        if self.sess().target.target.arch == "msp430" {
+            name = str::replace(&name[..], "@", "_");
```

Comment:
The function could probably be refactored away, yes.

As to MAX_BASE and `@`: Using a base64 encoding for the counter was an attempt to keep IR small after people complaining that it was too big. An unnecessary optimization as it turned out because it made no real difference.